megatron/rl/rl_utils.py

```diff
@@ -1326,751 +838,235 @@ def prepare_data_for_update(
-            # We run the above to fill in the inference/train side mismatch stats.
-            # We do the above for logging purposes.
-            # Nullify logprobs if not used in IS correction,
-            if not args.rl_inference_logprobs_is_correction:
-                inference_logprobs = None
-        elif not args.rl_use_sequence_packing:
-            # For sequence packing, inference_logprobs will be handled separately
-            inference_logprobs = None
-
-        # Handle packing of inference_logprobs for sequence packing mode
-        if (
-            args.rl_use_sequence_packing
-            and inference_logprobs is not None
-            and args.rl_inference_logprobs_is_correction
-        ):
-            with nvtx_range("pack-inference-logprobs"):
-                # Pack the inference logprobs using the helper function
-                packed_inference_logprobs = pack_inference_logprobs(
-                    inference_logprobs=inference_logprobs,
-                    packing_info=packing_context['packing_info'],
-                    generation_masks=generation_masks,
-                    bin_size=args.rl_sequence_packing_bin_size,
+            # Wrap forward_backward_func for Full iteration CUDA graph
+            forward_backward_func = get_forward_backward_func()
+            if args.enable_cuda_graph and args.cuda_graph_scope == "full_iteration":
+                forward_backward_func = FullCudaGraphWrapper(
+                    forward_backward_func, cuda_graph_warmup_steps=args.cuda_graph_warmup_steps
                 )
 
-                # Store packed inference logprobs in packing context
-                packing_context['packed_inference_logprobs'] = packed_inference_logprobs.cuda()
-                packing_context['has_inference_logprobs'] = True
-
-        with torch.no_grad(), nvtx_range("compute_ref_logprobs"):
-            # We need to load the ref model state dict and compute the logprobs for the ref model
-            cur_st_dict = {
-                k: (v.cpu() if v is not None else v) for k, v in model.state_dict().items()
-            }
-            model.load_state_dict(ref_state_dict)
-            ref_logprobs = []
-
-            # Compute reference logprobs
-            for batch_idx, (b_trajs, b_posids) in enumerate(data_iter):
-                # Get attention mask slice
-                if compute_attention_mask is not None:
-                    start_idx = batch_idx * logprobs_batch_size
-                    end_idx = min(
-                        start_idx + logprobs_batch_size, compute_attention_mask.shape[0]
+            def logprobs_forward_step(data_iterator, model):
+                if args.rl_use_sequence_packing:
+                    # When using sequence packing, the data iterator returns a tuple with a single element, the bin index.
+                    bin_tensor = next(data_iterator)[0]
+                    (b_trajs, _, _, _, b_posids, _, _, _, _, _, b_packed_seq_params) = (
```

Comment:
This becomes more and more dangerous. Can you add a todo to return a named tuple instead?